ddl/ddl_api.go

```diff
@@ -3292,19 +3276,78 @@ func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, keyType ast.Inde
 	return errors.Trace(err)
 }
 
-func buildFKInfo(fkName model.CIStr, keys []*ast.IndexColName, refer *ast.ReferenceDef, cols []*table.Column) (*model.FKInfo, error) {
+func buildFKInfo(fkName model.CIStr, keys []*ast.IndexColName, refer *ast.ReferenceDef, cols []*table.Column, tbInfo *model.TableInfo) (*model.FKInfo, error) {
 	var fkInfo model.FKInfo
 	fkInfo.Name = fkName
 	fkInfo.RefTable = refer.Table.Name
 
+	// all base columns of stored generated columns
+	baseCols := make(map[string]struct{})
+	for _, col := range cols {
+		if col.IsGenerated() && col.GeneratedStored {
+			for name := range col.Dependences {
+				baseCols[name] = struct{}{}
+			}
+		}
+	}
+
+	if len(keys) == 0 {
```

Comment:
The `baseCols` related logic is redundant if `len(keys) == 0` is satisfied.
The optimized arrange of this logic should be:
```go
// if len(keys) == 0 {}
// construct baseCols
// construct fkInfo
```

Reply:
And the better way to construct `fkInfo` is:
```go
fkInfo := &model.FKInfo {
	Name: fkName,
	RefTable: refer.Table.Name,
	Cols: make([]model.CIStr, len(keys)),
}
```
instead of allocating a object and assigning value for its fields.